### ISSUE TITLE
Support non-UTF-8 character encodings for HTML

### ISSUE THREAD

**Issue by k-bx** [servo/servo#6414]
Popular Ukrainian news website is rendered poorly on my Mac. All cyrillic letters are rendered as squares.

Steps to reproduce:

```
./mach run --release http://pravda.com.ua
```

Thanks for looking at it!

**Comment by k-bx** [comment]
Sorry, I see it uses windows-1251 encoding like it's a stone age. I'll write them a letter.

**Comment by SimonSapin** [comment]
That’s probably the issue, but we still need to fix it :)

Relevant spec: https://html.spec.whatwg.org/multipage/#determining-the-character-encoding
Related html5ever issue: https://github.com/servo/html5ever/issues/18

**Comment by vectorijk** [comment]
@SimonSapin It seems like Chinese, Japanese characters didn't be rendered properly on linux. Do you have idea or suggestion how to get involved and work on this issue? (lots of stuff seem need to be solved). Is this the encoding or rendering problem?

**Comment by Ms2ger** [comment]
Looks like there's a rendering problem too; that'd be another issue.

**Comment by SimonSapin** [comment]
Right, I think we also have bugs in font selection and font rendering, but I don’t know in details right now. Make sure to use UTF-8 for your HTML when testing to isolate from this bug.

**Comment by nhirata** [comment]
Chinese and Japanese characters are double byte and use UTF-16.

**Comment by SimonSapin** [comment]
@nhirata Many encodings other than UTF-16 support Chinese and Japanese characters, and “double byte character” is a term typically used in the context of the Shift-JIS encoding. This issue is about supporting both of them and more (https://encoding.spec.whatwg.org/#names-and-labels) through https://github.com/lifthrasiir/rust-encoding

There is already some support in html5ever (https://github.com/servo/html5ever/commit/2f4f64bbba1dfa4d5bd410260b90de0ffe7846a3 / https://github.com/servo/html5ever/pull/188) but we still need Servo to use it. I’m looking at this now.

**Comment by jdm** [comment]
Incomplete attempt to solve this: #9677

**Comment by SimonSapin** [comment]
With #9677 there would be two remaining pieces:

* In Servo, get the transport’s encoding information (`charset` in `Content-Type`) and communicate it to html5ever. (See [Kuchiki’s `from_hyper`](https://github.com/SimonSapin/kuchiki/blob/2e6d261cb39e555afad54d5cccdc9170df747cf9/src/parser.rs#L71-L75) for how to get this from an `hyper::client::Response`.)
* In html5ever, implement the rest of the encoding sniffing algorithm, in particular scan for `<meta>` elements: https://github.com/servo/html5ever/issues/18#issuecomment-187784806 (This may require refactoring how the parser is interrupted by blocking scripts, see https://github.com/servo/servo/pull/9677#issuecomment-185774132)

**Comment by jdm** [comment]
Latest attempt was #9730, but there were test failures that need further investigation.

**Comment by jdm** [comment]
For future reference, https://github.com/servo/html5ever/commit/79ed3e3207cf8578e9410b4310d67ea5bb06d990 and https://github.com/servo/servo/pull/16989 are related work that has happened. It would be good to figure out what work still remains to hook this up.

**Comment by shinglyu** [comment]
Looks like most previous work were on html5ever parsing. Are there related bugs in the font rendering module as well?

**Comment by SimonSapin** [comment]
This issue is about character encodings: converting bytes from the network to Unicode code points. I believe we also have font rendering issues but they’re independent from this and I don’t know if they’re filed.

**Comment by shinglyu** [comment]
I see. I did a quick search and found that there are several bugs here and there. Let me open a meta bug to collect them.

**Comment by nox** [comment]
We know read encoding from the HTTP `Content-Type` header.

**Comment by webbedspace** [comment]
This still affects various sites, including ones that use Shift-JIS.
https://intsys.co.jp/game/panepon/p01/index.html
Firefox:
<img width="485" height="365" alt="Image" src="https://github.com/user-attachments/assets/7df4de78-6d05-4db7-9568-0d3eac4d306c" />
Servo:
<img width="665" height="388" alt="Image" src="https://github.com/user-attachments/assets/ed35f132-4722-4091-b4a2-e7cf9aa9a58e" />
